crates/neon/src/types_impl/extract/mod.rs

```diff
@@ -0,0 +1,229 @@
+//! fn add(mut cx: FunctionContext) -> JsResult<JsNumber> {
+//!     let (a, b): (f64, f64) = cx.args()?;
+//!
+//!     Ok(cx.number(a + b))
+//! }
+//! ```
+//!
+//! ## Extracting [`Option`]
+//!
+//! It's also possible to mix [`Handle`], Rust types, and even [`Option`] for
+//! handling `null` and `undefined`.
+//!
+//! ```
+//! # use neon::{prelude::*, types::extract::*};
+//! fn get_or_default(mut cx: FunctionContext) -> JsResult<JsValue> {
+//!     let (n, default_value): (Option<f64>, Handle<JsValue>) = cx.args()?;
+//!
+//!     if let Some(n) = n {
+//!         return Ok(cx.number(n).upcast());
+//!     }
+//!
+//!     Ok(default_value)
+//! }
+//! ```
+//!
+//! ## Additional Extractors
+//!
+//! In some cases, the expected JavaScript type is ambiguous. For example, when
+//! trying to extract an [`f64`], the argument may be a `Date` instead of a `number`.
+//! Newtype extractors are provided to help.
+//!
+//! ```
+//! # use neon::{prelude::*, types::extract::*};
+//! # #[cfg(feature = "napi-5")]
+//! # use neon::types::JsDate;
+//!
+//! # #[cfg(feature = "napi-5")]
+//! fn add_hours(mut cx: FunctionContext) -> JsResult<JsDate> {
+//!     const MS_PER_HOUR: f64 = 60.0 * 60.0 * 1000.0;
+//!
+//!     let (Date(date), hours): (Date, f64) = cx.args()?;
+//!     let date = date + hours * MS_PER_HOUR;
+//!
+//!     cx.date(date).or_throw(&mut cx)
+//! }
+//! ```
+//!
+//! ## Overloaded Functions
+//!
+//! It's common in JavaScript to overload function signatures. This can be implemented with
+//! [`FunctionContext::args_opt`] or [`Context::try_catch`].
+//!
+//! ```
+//! # use neon::{prelude::*, types::extract::*};
+//!
+//! fn add(mut cx: FunctionContext, a: f64, b: f64) -> Handle<JsNumber> {
+//!     cx.number(a + b)
+//! }
+//!
+//! fn concat(mut cx: FunctionContext, a: String, b: String) -> Handle<JsString> {
+//!     cx.string(a + &b)
+//! }
+//!
+//! fn combine(mut cx: FunctionContext) -> JsResult<JsValue> {
+//!     if let Some((a, b)) = cx.args_opt()? {
+//!         return Ok(add(cx, a, b).upcast());
+//!     }
+//!
+//!     let (a, b) = cx.args()?;
+//!
+//!     Ok(concat(cx, a, b).upcast())
+//! }
+//! ```
+//!
+//! Note well, in this example, type annotations are not required on the tuple because
+//! Rust is able to infer it from the type arguments on `add` and `concat`.
+
+use crate::{
+    context::{Context, FunctionContext},
+    handle::Handle,
+    result::NeonResult,
+    types::JsValue,
+};
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+pub use self::json::*;
+pub use self::types::*;
+
+#[cfg(feature = "serde")]
+mod json;
+mod types;
+
+/// Extract Rust data from a JavaScript value
+pub trait TryFromJs<'cx>
+where
+    Self: private::Sealed + Sized,
+{
+    type Error;
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>;
+
+    fn from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Self>
```

Comment:
Do we not want a default implementation that calls `self.try_from_js()` and throws an exception on `Ok(Err())`?

Reply:
Currently, there cannot be a [useful] default implementation because there are no trait bounds on `type Error`. Currently the `TryFromJs` trait is sealed and cannot be implemented by users.

I opened an issue to stabilize `TryFromJs` and added this question. https://github.com/neon-bindings/neon/issues/1026